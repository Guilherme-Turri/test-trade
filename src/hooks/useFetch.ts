import React from 'react'

export const useFetchData = () => {
  const [error, setError] = React.useState<string | null>(null);
  const [loading, setLoading] = React.useState<boolean>(false);

  const request = React.useCallback(async <T> (url: string, options: RequestInit) => {
    let response;
    let json:T | null;
    setLoading(true)
    try {
      setError(null)
      response = await fetch(url, options)
      json = await response.json()
      if (!response.ok) throw new Error()
    } catch (e: any) {
      json = null;
      setError('Falha ao conectar com o servidor.')
    }
    finally {
      setLoading(false)
          }
    return { response, json } 
  }, []);
  return { error, request, loading }
} 