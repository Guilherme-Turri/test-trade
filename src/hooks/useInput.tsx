import React from 'react'

const typeField = {
  email: {
    regex: /^.{2,}$/,
    message: "Por favor, digite duas letras ou mais."
  }
}

export const useInput = () => {
  const [value, setValue] = React.useState<string>('')
  const [error, setError] = React.useState<string>('');

  const validate = (value: string) => {
    if (typeField.email.regex.test(value) === false) setError(typeField.email.message)
    else setError('');
  }

  const onChange = ({ target }: React.ChangeEvent<HTMLInputElement>) => {
    setValue(target.value)
    if (target.placeholder === "Nome") {
      validate(target.value)
    }
  }

  return (
    { value, onChange, error, setValue }
  )
}