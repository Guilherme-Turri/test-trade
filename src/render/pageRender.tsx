import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/theme';
import store from '../storage/store';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

export const pageRender = (children: React.ReactNode) => {
  return render(
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <BrowserRouter>
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>);
};