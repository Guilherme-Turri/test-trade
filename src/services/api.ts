const API_URL = ' https://v3.football.api-sports.io/'

export function AUTH_USER_API(apiKey:string) {
  return {
    url: `${API_URL}countries`,
    options: {
      method: 'GET',
      headers: {
      'x-rapidapi-host': 'v3.football.api-sports.io',
      'x-rapidapi-key': apiKey
    },
    },
  };
}
export function SEASON_API (apiKey:string) {
  return {
    url: `${API_URL}leagues/seasons`,
    options: {
      method: 'GET',
      headers: {
      'x-rapidapi-host': 'v3.football.api-sports.io',
      'x-rapidapi-key': apiKey
    },
    },
  };
}
export function LEAGUES_API (apiKey:string, country:string, season:string) {
    return {
    url: `${API_URL}leagues?country=${country}&season=${season}`,
    options: {
      method: 'GET',
      headers: {
      'x-rapidapi-host': 'v3.football.api-sports.io',
      'x-rapidapi-key': apiKey
    },
    },
  };
}
export function CLUB_API (apiKey:string, league:string, season:string) {
  return {
    url: `${API_URL}teams?league=${league}&season=${season}`,
    options: {
      method: 'GET',
      headers: {
      'x-rapidapi-host': 'v3.football.api-sports.io',
      'x-rapidapi-key': apiKey
    },
    },
  };
}
export function STATISTICS_API (apiKey:string, league:string, season:string, club:string) {
    return {
    url: `${API_URL}teams/statistics?league=${league}&season=${season}&team=${club}`,
    options: {
      method: 'GET',
      headers: {
      'x-rapidapi-host': 'v3.football.api-sports.io',
      'x-rapidapi-key': apiKey
    },
    },
  };
}

export function CHART_STATISTICS_API (apiKey:string, league:string, season:string, club:string) {
  return {
  url: `${API_URL}players?league=${league}&season=${season}&team=${club}`,
  options: {
    method: 'GET',
    headers: {
    'x-rapidapi-host': 'v3.football.api-sports.io',
    'x-rapidapi-key': apiKey
  },
  },
};
}