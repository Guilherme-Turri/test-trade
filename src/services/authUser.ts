import { AUTH_USER_API } from "./api";
import { IRequest, ICountry } from "../types/types";

interface AuthCountries{
  response?:ICountry[]
  errors?:{
    token:string
  }
}

export const authUser = async (apiKey:string , request:IRequest<AuthCountries>)=> {
  const { url, options } = AUTH_USER_API(apiKey)
  const {response, json} = await request(url, options)
  if(response && json){
    return {response, json}
   }
    else return 
 }
 