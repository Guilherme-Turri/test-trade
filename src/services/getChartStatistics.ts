import { CHART_STATISTICS_API } from "./api";
import { IRequest, IPlayer } from "../types/types";

interface PlayerProps{
  response?:IPlayer[],
  errors?:{
    token:string
  }
} 

export const getChartStatistics = async (apiKey:string, league:string, season:string, club:string, request:IRequest<PlayerProps>)=> {
    const { url, options } = CHART_STATISTICS_API(apiKey, league, season, club)
  const {response, json} = await request(url, options)
    if(response && json){
    return {response, json}
   }
    else return 
 }