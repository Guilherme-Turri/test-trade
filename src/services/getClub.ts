import { CLUB_API } from "./api";
import { IRequest, IClub } from "../types/types";

interface ClubPros{
  response?:IClub[],
  errors?:{
    token:string
  }
} 

export const getClub = async (apiKey:string, league:string, season:string, request:IRequest<ClubPros>)=> {
  const { url, options } = CLUB_API(apiKey, league, season)
  const {response, json} = await request(url, options)
  if(response && json){
    return {response, json}
   }
    else return 
 }