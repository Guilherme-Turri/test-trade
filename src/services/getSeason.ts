import { SEASON_API } from "./api";
import { IRequest } from "../types/types";

interface SeasonsProps{
  response?:string[],
  errors?:{
    token:string
  }
} 

export const getSeason = async (apiKey:string , request:IRequest<SeasonsProps>)=> {
  const { url, options } = SEASON_API(apiKey)
  const {response, json} = await request(url, options)
  if(response && json){
    return {response, json}
   }
    else return 
 }