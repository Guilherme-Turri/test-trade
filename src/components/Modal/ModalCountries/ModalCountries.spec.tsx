/* eslint-disable testing-library/no-unnecessary-act */
import { pageRender } from "../../../render/pageRender";
import { fireEvent, screen, waitFor, act } from '@testing-library/react'
import { ModalCountries } from ".";
import * as reactRedux from 'react-redux'
import * as getSeasonTest from '../../../services/getSeason';
import * as fetchDataTest from '../../../hooks/useFetch'

jest.setTimeout(10000)

const mockedUseDispatch = jest.fn()
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: () => mockedUseDispatch
}));

const allCountriesMock = [
  {
    code: 'BR',
    flag: 'www.flag/BR'
  },
  {
    code: 'ES',
    flag: 'www.flag/ES'
  },
]

const user = {
  apiKey: '123'
}

describe('<ModalCountries/>', () => {
  const useSelectorMock = reactRedux.useSelector as jest.Mock;
  const getSeasonspy = jest.spyOn(getSeasonTest, 'getSeason');

  it('should render on screen properly!', () => {
    useSelectorMock.mockReturnValue({
      parameters: {
        country: {
          all: allCountriesMock
        }
      }
    });

    const { container } = pageRender(<ModalCountries />)
    const item = screen.getByText('BR');
    expect(item).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  })

  it('Should render on the screen with a message to redo the search.', () => {
    useSelectorMock.mockReturnValue({
      parameters: null
    });
    pageRender(<ModalCountries />)
    const item = screen.getByText('Refaça a busca e/ou verifique as requisições diárias');
    expect(item).toBeInTheDocument();
  })

  it('should dispatch selected country into context', () => {
    useSelectorMock.mockReturnValue({
      user,
      parameters: {
        country: {
          all: allCountriesMock
        }
      }
    });

    pageRender(<ModalCountries />)

    const selectedCountry = screen.getByText('BR');
    act(() => {
      fireEvent.click(selectedCountry);
    });

    expect(mockedUseDispatch).toHaveBeenCalledWith({
      "payload": {
        "code": "BR",
        "flag": 'www.flag/BR'
      }, "type": "parameters/setCountry"
    });

  })
  it('should fetch all seasons from api and store on context', async () => {
    useSelectorMock.mockReturnValue({
      user,
      parameters: {
        country: {
          all: allCountriesMock
        }
      }
    });

    getSeasonspy.mockResolvedValue({
      response: { status: 200 },
      json: {
        response: ['2010', '2011', '2012']
      }
    });

    pageRender(<ModalCountries />)
    const selectedCountry = screen.getByText('BR');
    act(() => {
      fireEvent.click(selectedCountry);
    });

    expect(mockedUseDispatch).toHaveBeenCalledWith({
      "payload": {
        "code": "BR",
        "flag": 'www.flag/BR'
      }, "type": "parameters/setCountry"
    });

    await waitFor(() => {
      expect(mockedUseDispatch).toHaveBeenNthCalledWith(2, {
        "payload": ['2010', '2011', '2012'],
        "type": "parameters/setSeasons"
      });
    })
  })

  it('should display error message on getSeason', async () => {
    useSelectorMock.mockReturnValue({
      user,
      parameters: {
        country: {
          all: allCountriesMock
        }
      }
    });

    getSeasonspy.mockResolvedValue({
      response: { status: 200 },
      json: {
        response: [],
        errors: { token: 'Verifique novamente a busca' }
      }
    });

    pageRender(<ModalCountries />)
    const selectedCountry = screen.getByText('BR');
    act(() => {
      fireEvent.click(selectedCountry);
    });

    expect(mockedUseDispatch).toHaveBeenCalledWith({
      "payload": {
        "code": "BR",
        "flag": 'www.flag/BR'
      }, "type": "parameters/setCountry"
    });

    await waitFor(() => {
      const msg = screen.getByText('Verifique novamente a busca')
      expect(msg).toBeInTheDocument();
    })
  })
})

describe('<ModalCountries/> -  display right actions by the custom hook useFetch', () => {

  const useSelectorMock = reactRedux.useSelector as jest.Mock;

  it('should display loading on screen', async () => {
    jest.spyOn(fetchDataTest, "useFetchData").mockReturnValue({
      error: null,
      request: jest.fn(),
      loading: true
    });

    useSelectorMock.mockReturnValue({
      user,
      parameters: {
        country: {
          all: allCountriesMock
        }
      }
    });

    pageRender(<ModalCountries />)

    await waitFor(() => {
      expect(screen.getByTestId('Ball')).toBeInTheDocument();
    })
  })

  it('should display errormsg on screen', async () => {
    jest.spyOn(fetchDataTest, "useFetchData").mockReturnValue({
      error: 'ErrorMsg',
      request: jest.fn(),
      loading: false
    });
    useSelectorMock.mockReturnValue({
      user,
      parameters: {
        country: {
          all: allCountriesMock
        }
      }
    });

    pageRender(<ModalCountries />)
    const selectedCountry = screen.getByText('BR');
    act(() => {
      fireEvent.click(selectedCountry);
    });
    await waitFor(() => {
      expect(screen.getByText('ErrorMsg')).toBeInTheDocument();
    })
  })
})   