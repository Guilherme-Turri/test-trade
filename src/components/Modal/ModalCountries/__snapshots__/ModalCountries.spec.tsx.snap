// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalCountries/> should render on screen properly! 1`] = `
.c0 {
  height: 100%;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.938rem;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  border-radius: 0.625rem;
  overflow-y: auto;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  cursor: pointer;
  width: 4.688rem;
  height: 3.125rem;
}

.c3 {
  width: 1.563rem;
  height: 1.563rem;
  border-radius: 50%;
}

.c2 {
  color: #2d5839;
}

<div>
  <div
    class="c0"
    data-testid="countriesContainer"
  >
    <div
      class="c1"
      data-testid="container"
    >
      <p
        class="c2"
      >
        BR
      </p>
      <img
        alt="BR"
        class="c3"
        src="www.flag/BR"
      />
    </div>
    <div
      class="c1"
      data-testid="container"
    >
      <p
        class="c2"
      >
        ES
      </p>
      <img
        alt="ES"
        class="c3"
        src="www.flag/ES"
      />
    </div>
  </div>
</div>
`;
